### PR TITLE
alpine linux compatibility fixes

### LINKED ISSUE #860: Atomic operations detections fails on OpenWRT/LEDE
https://github.com/firehol/netdata/commit/db787ff47f80740ae266e0e347a6b49b3a395c35

After this commit netdata fails to compile using the LEDE-project toolchain.

`global_statistics.o: In function `finished_web_request_statistics':
global_statistics.c:(.text+0x90): undefined reference to `__atomic_compare_exchange_8'
global_statistics.c:(.text+0xc8): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0xe0): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0xf8): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0x110): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0x128): undefined reference to `__atomic_fetch_add_8'
global_statistics.o:global_statistics.c:(.text+0x140): more undefined references to `__atomic_fetch_add_8' follow
global_statistics.o: In function `global_statistics_copy':
global_statistics.c:(.text+0x3a4): undefined reference to `__atomic_compare_exchange_8'
collect2: error: ld returned 1 exit status
Makefile:575: recipe for target 'netdata' failed
make[6]: *** [netdata] Error 1
`

### PR DESCRIPTION
**Pull request by ktsaou**
1. fixes all warnings / errors when compiling against musl
2. fixes all other issues detected when the system is busybox based
3. fixes #860

### CODE REVIEW

**Comment by ktsaou** [comment]
@alonbl I have made something here I don't like. I check for the C11 `_Generic` directive using the gcc version.

I see that there is a solution for autoconf: https://lists.gnu.org/archive/html/autoconf-patches/2012-12/msg00017.html

Can you help me? I remind you that we require autotools >2.60, so I don't know if this is already included or we have to somehow add this code to `configure.ac`.

File: src/common.h [hunk 1]
```diff
@@ -65,9 +65,11 @@
 #endif
 
 #ifdef __GNUC__
-#define GCC_VERSION (__GNUC__ * 10000 \
-                               + __GNUC_MINOR__ * 100 \
-                               + __GNUC_PATCHLEVEL__)
+#define GCC_VERSION (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__)
+
+#if GCC_VERSION < 40900
+#define WITHOUT_C11_GENERIC 1
```
Comment (ktsaou) [inline src/common.h]: @alonbl the check is here

File: src/log.c [hunk 2]
```diff
@@ -266,6 +266,30 @@ void info_int( const char *file, const char *function, const unsigned long line,
 // ----------------------------------------------------------------------------
 // error log
 
+#ifdef WITHOUT_C11_GENERIC
```
Comment (ktsaou) [inline src/log.c]: @alonbl this is where it is used.

What I need is a solution for `strerror_r`. The glibc version returns `char *`, while all others return `int`. The problem is that the glibc version might not use the `buf` supplied at all. So, we really need to know which version the current system has.

File: src/log.c [hunk 3]
```diff
@@ -266,6 +266,30 @@ void info_int( const char *file, const char *function, const unsigned long line,
...
+// http://stackoverflow.com/questions/479207/function-overloading-in-c
+static const char *strerror_result_int(int a, const char *b) { (void)a; return b; }
+static const char *strerror_result_string(const char *a, const char *b) { (void)b; return a; }
+
+#define strerror_result(a, b) _Generic((a), \
+    int: strerror_result_int, \
+    char *: strerror_result_string \
+    )(a, b)
+
+#endif /* ! WITHOUT_C11_GENERIC */
```
Comment (ktsaou) [inline src/log.c]: this is the elegant C11 solution....

**Comment by ktsaou** [comment]
@alonbl I improvised a bit with autoconf to detect the availability of atomic operations and detecting strerror_r using autoconf.

If you have any comments, please let me know.
Thanks!

**Comment by alonbl** [comment]
Looks nice!
You should now go into removing remaining arch/build detection and conditionals and move them to autoconf as well.
For example after you use AC_CANONICAL_HOST you can use a simple case to extract the arch. You can also use AC_CHECK_SIZEOF to determine void* size to detect 32/64 arch.
As you have done this nicely using macros, I suggest you put each in own m4/xxx.m4 file, these will be included when you autoreconf without need to bloat the configure.ac.

**Pull request merged**